### PR TITLE
Add BSI based on Roaring64Bitmap

### PR DESCRIPTION
**Pull request by iszhangpch**
### SUMMARY
- Add BSI based on Roaring64Bitmap so as to support (column id:bigint, value:bigint)

### Automated Checks

- [x] I have run `./gradlew test` and made sure that my PR does not break any unit test.
- [x] I have run `./gradlew checkstyleMain` or the equivalent and corrected the formatting warnings reported.

### CODE REVIEW

**Comment by lemire** [comment]
Looks reasonable at a glance.

File: RoaringBitmap/src/main/java/org/roaringbitmap/longlong/Roaring64Bitmap.java [hunk 1]
```diff
@@ -334,6 +334,40 @@ public void or(final Roaring64Bitmap x2) {
...
 
+  /**
+   * Bitwise OR (union) operation. The provided bitmaps are *not* modified. This operation is
+   * thread-safe as long as the provided bitmaps remain unchanged.
+   *
+   * @param x1 first bitmap
+   * @param x2 other bitmap
+   * @return result of the operation
+   */
+  public static Roaring64Bitmap or(final Roaring64Bitmap x1, final Roaring64Bitmap x2) {
```
Comment (blacelle) [inline RoaringBitmap/src/main/java/org/roaringbitmap/longlong/Roaring64Bitmap.java:345]: Should the implementation be subsentially more similar to `https://github.com/RoaringBitmap/RoaringBitmap/blob/master/RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java#L2471` ?

Comment (iszhangpch) [reply RoaringBitmap/src/main/java/org/roaringbitmap/longlong/Roaring64Bitmap.java:345]: Yes, I agree. It's been modified.

**Review by blacelle — APPROVED** [review]
LGTM

**Comment by richardstartin** [comment]
I haven't really scrutinised the code much but I'm curious about the use case. We could easily extend `RangeBitmap` to 64 bit row indices, and the savings in row index prefix metadata could be significant. Is mutability important?

**Comment by iszhangpch** [comment]
@richardstartin Thank you so much for your question.

Our use cases are mainly user protrait analysis scenarios, Roaring Bitmap can transform relational operations into intersection and difference operations to improve performance, but we will encounter the following problems:

Multi-tag joint query problem:
- Roaring Bitmap is more often used for "attribute labels", but for "behavioral labels", such as PV, order amount and other quantitative value labels, can not be well applied.

The query problem of high cardinality tags:
- When the number of de-duplicated values (Cardinality) of a tag is very large, the storage of Roaring Bitmap will be inflated, resulting in query performance degradation.

Therefore, we need to optimize by Bit-sliced Index and several functions.

For example:
| cid | value (D) | value (B) |
|:------|:------|:------|
| 1 | 3 | 0011 |
| 2 | 6 | 0110|
| 3 | 4 | 0100 |
| 4 | 10 | 1010|
| 5 | 7 | 0111 |

Then, we can get the Bit-sliced Index:
- slice 0: roaringbitmap '{1, 5}'
- slice 1: roaringbitmap '{1, 2, 4, 5}'
- slice 2: roaringbitmap '{2, 3, 5}'
- slice 3: roaringbitmap '{4}'

If we already have the crowd= '{1,3,5}' that we need from the bitmap calculation of RoaringBitmap, we can further compute it efficiently through BSI, such as, to compute the sum of values corresponding to the crowd:
$$
rb\_and\_cardinality(crowd, slice_0) * 2^0 + ... + rb\_and\_cardinality(crowd, slice_3) * 2^3 = 14
$$

In short, with BSI, we can not only compress the key-value data, but also combine the RoaringBitmap calculation of cid with the value calculation.

For RangeBitmap I know relatively limited, I understand that it is also a very good data structure that can be used for data compression and range query, but maybe not quite suitable for our scenario, where mutability is not a priority.

I hope my answer can solve your problem, if there is something unprofessional in my answer, you are welcome to guide me!

**Comment by richardstartin** [comment]
thanks for explaining. RangeBitmap is also an immutable bit sliced index for range and equality queries over high cardinality attributes (it takes any 64 bit value), it’s just limited to 2^32 rows. There are some details here: https://richardstartin.github.io/posts/range-bitmap-index

I’m fine with merging this, just trying to understand the gap.

**Comment by iszhangpch** [comment]
@richardstartin Your documentation is much appreciated, I have read the documentation you provided as well as the implementation of RangeBitmap within the project. I have understood that RangeBitmap is also based on the same idea of bit slicing to build range encoding.

Our main requirement lies in the fact that apart from the query functions such as EQ, GE, GT, LE, NEQ, etc., we also need to perform some numerical computation of aggregation analysis functions such as sum and topK against value via BitmapSlideIndex, which we are currently implementing in the UDF of the OLAP system.

I think this is the main gap.

**Comment by lemire** [comment]
Though there is no objection to this PR, I want to leave it open deliberately a little bit longer so that discussions can happen and any misunderstanding can be resolved.

**Comment by lemire** [comment]
I am deliberately leaving this open, inviting further discussions, since this would extend our public API.

**Comment by iszhangpch** [comment]
@lemire I found an issue that may cause RoaringBSI's fountSet to not take effect correctly when doing LE and GE comparisons, the 64-bit RoaringBSI's fix I submitted in this PR, and the 32-bit BSI's fix I'm going to open a separate PR.

**Comment by iszhangpch** [comment]
> I am deliberately leaving this open, inviting further discussions, since this would extend our public API.

Hello, Lemire, Can this PR be merged? It's been a long time.

**Comment by lemire** [comment]
Ok. Let us merge. This will be part of the next release.

**Comment by Smallhi** [comment]
@iszhangpch For bsi, I think rb32 is better.  since rb64 may cause OOM when aggregation.
in our case , we bucket by int(cid/bucket_size) ( bucket_size=1_000_000)  and use two stage compution method. 
eg:
select sum(pv) as pv
(
select bucket_num,bsi_sum(bsi_data, foundset) as pv  from data 
group by bucket_num
) t
